packages/flutter_tools/lib/src/android/android_console.dart

```diff
@@ -0,0 +1,50 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:async/async.dart';
+
+/// Creates a console connection to an Android emulator that can be used to run
+/// commands such as "avd name" which are not available to ADB.
+class AndroidConsole {
```

Comment:
I think we should have tests for this. I would inject a mock'd socket and work from there

Reply:
I think I've finally done this, but it was more complicated than I expected so I don't know if I did it all right...

I needed to use Context in order to get a different one in the tests (a mock) but since we need (well, want) an instance-per-connection I had to make the context value a typedef/factory function. I couldn't find any existing code doing similar things to know if this is best (except VM WebSockets which are similar).

PTAL and let me know if anything seems weird and I'll fix/change. Thanks!